pkg/azure/client/storage.go

```diff
@@ -28,9 +30,46 @@ type BlobStorageClient struct {
 	client *azblob.Client
 }
 
-// newStorageClient creates a client for an Azure Blob storage by reading auth information from secret reference. Requires passing the storage domain (formerly
+// BlobStorageDomainFromCloudConfiguration returns the storage service domain given a known cloudConfiguration.
+func BlobStorageDomainFromCloudConfiguration(cloudConfiguration *azureapi.CloudConfiguration) (string, error) {
+	// Unfortunately the valid values for storage domains run by Microsoft do not seem to be part of any sdk module. They might be queryable from the cloud configuration,
+	// but I also haven't been able to find a documented list of proper ServiceName values.
+	// Furthermore, it seems there is still no unified way of specifying the cloud instance to connect to as the domain remains part of the storage account URL while
+	// the new options _also_ allow configuring the cloud instance.
+	switch {
+	case cloudConfiguration == nil:
+		return "blob.core.windows.net", nil
+	case strings.EqualFold(cloudConfiguration.Name, "AzurePublic"):
+		return "blob.core.windows.net", nil
+	case strings.EqualFold(cloudConfiguration.Name, "AzureGovernment"):
+		// Note: This differs from the one mentioned in the docs ("blob.core.govcloudapi.net") but should be the right one.
+		// ref.: https://github.com/google/go-cloud/blob/be1b4aee38955e1b8cd1c46f8f47fb6f9d820a9b/blob/azureblob/azureblob.go#L162
+		return "blob.core.usgovcloudapi.net", nil
+	case strings.EqualFold(cloudConfiguration.Name, "AzureChina"):
+		// source: https://learn.microsoft.com/en-us/azure/china/resources-developer-guide#check-endpoints-in-azure
+		return "blob.core.chinacloudapi.cn", nil
+	}
+	return "", fmt.Errorf("unknown cloud configuration name '%s'", cloudConfiguration.Name)
+}
+
+// NewStorageClient creates a blob storage client.
+func NewStorageClient(_ context.Context, storageAccountName, storageAccountKey, storageDomain string) (*BlobStorageClient, error) {
+	credentials, err := azblob.NewSharedKeyCredential(storageAccountName, storageAccountKey)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create shared key credentials: %v", err)
+	}
+
+	storageEndpointURL, err := url.Parse(fmt.Sprintf("https://%s.%s", storageAccountName, storageDomain))
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse service url: %v", err)
+	}
+	blobclient, err := azblob.NewClientWithSharedKeyCredential(storageEndpointURL.String(), credentials, nil)
```

Comment:
If you look at the azure-sdk-for-go the nomenclature is a bit different. They provide 3 different clients AFAIK:

1. Blob client which is created with a specific blob URL (see [here](https://github.com/Azure/azure-sdk-for-go/blob/main/sdk/storage/azblob/blob/client.go#L38))
2. Container client which is created for a specific container/bucket ( see [here](https://github.com/Azure/azure-sdk-for-go/blob/main/sdk/storage/azblob/container/client.go#L44)).
3. Service client to manage buckets/containers (see [here](https://github.com/Azure/azure-sdk-for-go/blob/main/sdk/storage/azblob/service/client.go#L42))

So what you have here is a `ContainerClient` or a `BucketClient` but you call it a blobClient. This confused me a bit.